pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go

```diff
@@ -281,17 +281,23 @@ func (le *LeaderEngine) IsLeader() bool {
 	return le.GetLeader() == le.HolderIdentity
 }
 
-// Subscribe allows any component to receive a notification
-// when the current process becomes leader.
+// Subscribe allows any component to receive a notification when leadership state of the current
+// process changes.
+//
+// The subscriber will not be notified about the leadership state change if the previous notification
+// hasn't yet been consumed from the notification channel.
+//
 // Calling Subscribe is optional, use IsLeader if the client doesn't need an event-based approach.
-func (le *LeaderEngine) Subscribe() <-chan struct{} {
-	c := make(chan struct{}, 5) // buffered channel to avoid blocking in case of stuck subscriber
```

Comment:
Check the implementation of the notify function.

If the buffer is already full, no new notifications are placed on the channel. 

We only notify on the subscriber's channel if the channel is already empty, so we will never get stuck when writing to the channel.

Reply:
basically [this](https://github.com/DataDog/datadog-agent/blob/290d4bac7776f461dce431287925ab86fcd74aa0/pkg/util/kubernetes/apiserver/leaderelection/leaderelection_engine.go#L230-L233)

Reply:
Gotcha that makes sense. I'm trying to think of edge cases in leadership election (such as when leadership may be gained, lost, and then regained quickly) and how this might effect the overall state.

In such a case where
1. The Cluster Agent X is elected leader and sends a notification to it's subscribers via `notify()`
2. The subscriber does not get CPU time to process the notification.
3. The Cluster Agent X is then removed as leader and attempts to send a notification to it's subscribers via `notify()`
* However in this case, since the channel is full (buffer len = 1), it will not send a second notification.
4. The subscriber is finally scheduled on the CPU and only receives the first Leader Notification and doesn't know about the lost leadership notification.

The buffer length of 5 provided some resiliency in this type of situation I believe but also isn't perfect itself.

Reply:
Thanks for the explanation Adel about tracking _changes_.

So it seems in your implementation, if the leader cluster agent first sends a notification when it lost leadership, subscriber doesn't have time to read it, then sends another notification when it re-gains leadership, the subscriber may then get unblocked and process the notification but there has been no perceived change in leader. Whereas previously it would've been aware of the 2 changes in leadership that occurred via the buffer. I don't know if this slight change in behavior is relevant for consistency but it seems fine.

Overall it looks good to me if this isn't a concern.

Reply:
Previously, the notification was sent only when the instance becomes a leader. No notification was being sent in case of leadership loss. 

Therefore, in the scenario you described, in both implementations, only 1 notification is sent.

Perhaps what you wanted to say is that if the instance acquires leadership, loses it, then acquires leadership again, in the old implementation, this would've produced 2 notifications regardless of whether the subscriber consumed the first implementation or not. In the new implementation, this will ideally send at most 3 notifications assuming that the subscriber consumes each notification immediately. However, if the subscriber doesn't consume the first notification at all, the second and third notifications won't be sent. Then if after that the subscriber is able to consume the notification on the channel, it will then be able to check the `IsLeader` function.

It is important to note that the subscriber doesn't care about the history of changes in the leadership state. In practice this information is not useful in most cases. Subscribers are only interested in being notified when the state has changed. From the point of view of the subscriber, saying simultaneously that **leadership state changed once** or **leadership state changed N times** are essentially equivalent, because the subscriber doesn't care about the history of changes, but rather about the most recent state. 

If at some point in the future we have a use case where the subscriber is interested in knowing the number of changes that occurred in the past, we can modify the code to support that. For now, this use case doesn't exist and is not expected to be there any time soon.

I hope this makes sense.